### PR TITLE
Issue #255 Address structure of secure-session

### LINKED ISSUE #255: Structure of secure-session and secure-session-enable
We have the following structure in ietf-bgp.yang:

```
      leaf secure-session-enable {
            type boolean;
            default "false";
            description
              "Does this session need to be secured?";
          }

          container secure-session {
            /* XXX JMH - why is this constraint mandatory? also, why not have the enable in this container? */
            when "../secure-session-enable = 'true'";
            description
              "Container for describing how a particular BGP session
               is to be secured.";
```

As noted by my comment in secure-session, why is that constraint mandatory?
Also, why is the enable outside of the container?

I suspect we want to move secure-session-enable into the secure-session container as "enable".  This also permits configurable of security parameters without necessarily having them enabled at the moment, although some might consider that a bug rather than a feature.  (You configured it... did you even bother to turn it on?

### PR DESCRIPTION
**Pull request by jhaas-pfrc**
Issue #255 Address structure of secure-session
    
The enabled boolean didn't reside in the container with the secure-session
parameters themselves.  The constraint on the existing options had a constraint
that you couldn't configure options without setting the enabled status to true.
This is the opposite of how you'd normally want this to operate.
    
Move the enabled into the secure-session container.
Create a must constraint that when enabled, options MUST be configured.
To support the must constraint, put the current choice option into new container "options" to provide a XPath testable element.
Move secure-session into the transport container and remove redundant copy under peer-groups.
Fix the examples using secure-session.

Closes #255

### CODE REVIEW

**Comment by jhaas-pfrc** [comment]
> Noticed that you dropped IPSec as an authentication option. Is that on purpose?

That's an error in the change.  That's why we have reviews. :-)

**Comment by jhaas-pfrc** [comment]
> Noticed that you dropped IPSec as an authentication option. Is that on purpose?

Looking at the diff, the issue was that ipsec was not consistently present in both places we had secure-session.  So, an accidental bug fix is happening. :-)

**Comment by jhaas-pfrc** [comment]
@mjethanandani I believe I've addressed the open issues here. Could you verify?

File: src/yang/ietf-bgp-common.yang [hunk 1]
```diff
@@ -340,6 +345,82 @@ submodule ietf-bgp-common {
...
+          "Options for securing the BGP session.";
+        choice option {
+          case ao {
+            leaf ao-keychain {
+              type key-chain:key-chain-ref;
+              description
+                "Reference to the key chain that will be used by this
+                 model. Applicable for TCP-AO and TCP-MD5 only";
+              reference
+                "RFC 8177: YANG Key Chain.";
```
Comment (mjethanandani) [inline src/yang/ietf-bgp-common.yang:377]: The title of 8177 is different from the title above in the import statement.

File: src/yang/ietf-bgp-common.yang [hunk 2]
```diff
@@ -340,6 +345,82 @@ submodule ietf-bgp-common {
...
+          }
+
+          case md5 {
+            leaf md5-keychain {
+              type key-chain:key-chain-ref;
+              description
+                "Reference to the key chain that will be used by this
+                 model. Applicable for TCP-AO and TCP-MD5 only";
+              reference
+                "RFC 8177: YANG Key Chain.";
```
Comment (mjethanandani) [inline src/yang/ietf-bgp-common.yang:394]: Same here.

File: src/yang/ietf-bgp-common.yang [hunk 3]
```diff
@@ -340,6 +345,82 @@ submodule ietf-bgp-common {
...
+            }
+            description
+              "Uses TCP-AO to secure the session. Parameters for
+               those are defined as a grouping in the TCP YANG
+               model.";
+            reference
+              "RFC 5925 - The TCP Authentication Option.";
```
Comment (mjethanandani) [inline src/yang/ietf-bgp-common.yang:384]: While you are here, can you fix the format of this reference statement.

**Comment by mjethanandani** [comment]
Sorry for not catching some of this in the initial review.

File: src/yang/ietf-bgp-common.yang (hunk 1 above)
Comment (jhaas-pfrc) [reply src/yang/ietf-bgp-common.yang:377]: Fixed in this file.

File: src/yang/ietf-bgp-common.yang (hunk 2 above)
Comment (jhaas-pfrc) [reply src/yang/ietf-bgp-common.yang:394]: Fixed in this file.

File: src/yang/ietf-bgp-common.yang (hunk 3 above)
Comment (jhaas-pfrc) [reply src/yang/ietf-bgp-common.yang:384]: Fixed in this file.

File: src/yang/ietf-bgp-common.yang [hunk 4]
```diff
@@ -374,14 +374,14 @@ submodule ietf-bgp-common {
...
-                "RFC 8177: YANG Key Chain.";
+                "RFC 8177: YANG Data Model for Key Chains.";
             }
             description
               "Uses TCP-AO to secure the session. Parameters for
                those are defined as a grouping in the TCP YANG
                model.";
             reference
-              "RFC 5925 - The TCP Authentication Option.";
+              "RFC 5925:he TCP Authentication Option.";
```
Comment (mjethanandani) [inline src/yang/ietf-bgp-common.yang:384]: s/he/ The/

**Review by mjethanandani — APPROVED** [review]

**Pull request merged**